server/src/main/java/org/opensearch/action/search/SearchShardTask.java

```diff
@@ -38,15 +38,35 @@
 import org.opensearch.tasks.TaskId;
 
 import java.util.Map;
+import java.util.function.Supplier;
 
 /**
  * Task storing information about a currently running search shard request.
  * See {@link ShardSearchRequest}, {@link ShardFetchSearchRequest}, ...
  */
 public class SearchShardTask extends CancellableTask {
+    // generating metadata in a lazy way since source can be quite big
+    private final Supplier<String> metadataSupplier;
```

Comment:
Same/similar question from me - is there a reason why this can't be a simple `String` instead of a `Supplier` ?

Reply:
For `SearchShardTask` we want to log the query string to be able to correlate the top resource consumers with the query. We have seen query strings can be really huge and with high search workload it can be very expensive to construct it for each shard task. Also depending on the consumers type it may not be needed to get the query string for each shard tasks. Hence, depending upon the consumers (for example in this case `TopNConsumer`) it will only call this method on the Tasks which are of interest and not on all the Task object. 

I agree with the memoization approach but given at the moment there is only one caller expected for this method seems like we can do that optimization as a follow-up ?

Reply:
As @sohami pointed out this method will only be added once (from the consumer and if query is top resource consumer), there are no other callers to this method. In future if this method is going to be called more than once, we can extend to memoize the result of this supplier. Adding a comment for future reference.